### PR TITLE
add options to individually include hw mul and div/mod

### PR DESCRIPTION
**Pull request by vanjoe**
I added -mmul and -mdiv to the target specific options. I kept -mmuldiv around for legacy. 

If anyone knows a better way of making -mmuldiv simply expand to -mmul -mdiv, I'd like to know, otherwise i think my way works.

### CODE REVIEW

File: gcc/common/config/riscv/riscv-common.c [hunk 1]
```diff
@@ -110,9 +110,16 @@ riscv_handle_option (struct gcc_options *opts,
...
+	    opts->x_target_flags |= (MASK_MUL | MASK_DIV);
+	    //error("OPT_mmuldiv found");
+	    return true;
+    case OPT_mno_muldiv:
+	    opts->x_target_flags &= ~(MASK_MUL | MASK_DIV);
+	    //error("OPT_mno_muldiv found");
+	    return true;
     default:
-      return true;
+	    return true;
```
Comment (kito-cheng) [inline gcc/common/config/riscv/riscv-common.c]: Wrong indent and unclean up code.

File: gcc/config/riscv/riscv.opt [hunk 2]
```diff
@@ -71,9 +71,21 @@ Target Report Mask(ATOMIC)
 Use hardware atomic memory instructions.
 
 mmuldiv
-Target Report Mask(MULDIV)
+Target Report RejectNegative Var(TARGET_muldiv)
 Use hardware instructions for integer multiplication and division.
 
+mno-muldiv
+Target Report RejectNegative Undocumented Var(TARGET_no_muldiv)
+Dont Use hardware instructions for integer multiplication and division.
```
Comment (kito-cheng) [inline gcc/config/riscv/riscv.opt]: I think let remove  RejectNegative for mmuldiv is more reasonable since mno-muldiv is Undocumented, 
and Var(TARGET_muldiv) can remove if it's unused.

File: gcc/config/riscv/riscv.md [hunk 3]
```diff
@@ -169,7 +169,7 @@
 ;; D2S          double to float single
 ;; S2D          float single to double
 
-(define_attr "cnv_mode" "unknown,I2S,I2D,S2I,D2I,D2S,S2D" 
+(define_attr "cnv_mode" "unknown,I2S,I2D,S2I,D2I,D2S,S2D"
```
Comment (kito-cheng) [inline gcc/config/riscv/riscv.md]: Clean up should split into another patch instead of mix with feature implementation patch.

**Comment by kito-cheng** [comment]
btw, DRIVER_SELF_SPECS can handle option too.

for example:
```
diff --git a/gcc/config/riscv/riscv.h b/gcc/config/riscv/riscv.h
index 4a55ae2..78a848c 100644
--- a/gcc/config/riscv/riscv.h
+++ b/gcc/config/riscv/riscv.h
@@ -131,7 +131,9 @@ along with GCC; see the file COPYING3.  If not see
   {"tune", "%{!mtune=*:-mtune=%(VALUE)}" }, \
   {"float", "%{!msoft-float:%{!mhard-float:-m%(VALUE)-float}}" }, \
 
-#define DRIVER_SELF_SPECS ""
+#define DRIVER_SELF_SPECS \
+  " %{mno-muldiv:-mno-mul -mno-div} " \
+  " %{mmuldiv:-mmul -mdiv} "
 
 #ifdef IN_LIBGCC2
 #undef TARGET_64BIT

```

**Comment by aswaterman** [comment]
My preference would be to avoid DRIVER_SELF_SPECS for option handling.

**Comment by palmer-dabbelt** [comment]
This looks generally good.  @vanjoe: can you address Kito's comments?  We want to keep "-mno-muldiv" and "-mmuldiv", since other stuff has been using them.  You can just force push over this PR and it'll update.

@kito-cheng Thanks for looking at this!

**Comment by vanjoe** [comment]
Updated

File: gcc/common/config/riscv/riscv-common.c [hunk 4]
```diff
@@ -110,7 +110,13 @@ riscv_handle_option (struct gcc_options *opts,
...
       riscv_parse_arch_string (decoded->arg, &opts->x_target_flags);
       return true;
-
+    case OPT_mmuldiv:
+      if (decoded->value){
+      opts->x_target_flags |= (MASK_MUL | MASK_DIV);
+      }else{
+         opts->x_target_flags &= ~(MASK_MUL | MASK_DIV);
+      }
+      return true;
```
Comment (palmer-dabbelt) [inline gcc/common/config/riscv/riscv-common.c]: This doesn't match the GNU indenting style.

File: gcc/config/riscv/riscv.h [hunk 5]
```diff
@@ -63,7 +63,11 @@ along with GCC; see the file COPYING3.  If not see
       if (TARGET_ATOMIC)						\
 	builtin_define ("__riscv_atomic");				\
 									\
-      if (TARGET_MULDIV)						\
+      if (TARGET_MUL)						\
+	builtin_define ("__riscv_mul");				\
+      if (TARGET_DIV)						\
+	builtin_define ("__riscv_div");				\
+      if (TARGET_DIV && TARGET_MUL)						\
```
Comment (palmer-dabbelt) [inline gcc/config/riscv/riscv.h]: It's possible the patch has broken these indents, but I think the \'s don't line up.

Comment (vanjoe) [inline gcc/config/riscv/riscv.h]: None of the \'s line up in my emacs. Will fix.

What test suite do you want me to run?

Comment (palmer-dabbelt) [inline gcc/config/riscv/riscv.h]: The newlib test suite should be fine:

https://github.com/riscv/riscv-gnu-toolchain/blob/master/README.md#test-suite

File: gcc/common/config/riscv/riscv-common.c [hunk 6]
```diff
@@ -110,7 +110,16 @@ riscv_handle_option (struct gcc_options *opts,
...
+    case OPT_mmuldiv:
+      if (decoded->value)
+         {
+            opts->x_target_flags |= (MASK_MUL | MASK_DIV);
+         }
+      else
+         {
+            opts->x_target_flags &= ~(MASK_MUL | MASK_DIV);
+         }
+      return true;
```
Comment (kito-cheng) [inline gcc/common/config/riscv/riscv-common.c]: brackets can be removed if the `if` or `else` body contain only 1 line.

**Comment by vanjoe** [comment]
Tests report : 
		=== gcc Summary ===

# of expected passes		12790
# of unexpected failures	15
# of expected failures		7
# of unsupported tests		167


Is that normal?

**Comment by palmer-dabbelt** [comment]
It's not what mine looks like, but I probably screwed something up in my own directory.  This is merged

  https://github.com/riscv/riscv-gcc/commit/4aabfe51c4678948f02609d3a898c2a2c2df466f

but github isn't picking it up.

Thanks for the patches!
